### ISSUE TITLE
Improve docker image

### ISSUE THREAD

**Issue by elfjes** [pypiserver/pypiserver#359]
There are currently a number of issues with the docker image
 - Requires root permissions (which leads to #357)
 - Using the docker image for anything else than docker run (this currently also throws an error in the docker tests (when the script calls `docker run pypiserver:test --help` , but this error is ignored)
- It uses the default wsgiref server which is non-theaded and not production ready. We should pick another.
- I noticed that the pypiserver uses /data as a root, but /data/packages as a volume, which means that new package uploads may not be persisted
- Currently `PORT` is used as an environment variable to customize the port. While environment variables for configuration of containers is considered good practice, I believe we should make this variable more specifc, ie `PYPISERVER_PORT`. We can then also expand this to other configuration settings, and maybe not limit it to only the docker image.

Together with the lack of testing (#346) I think the docker image release of pypiserver should be given some more attention. I think the `entrypoint.sh` currently does more harm than good and we should try and remove it again (while keeping docker-compose working), or improve on it drastically

**Comment by mplanchard** [comment]
> Requires root permissions (which leads to #357)

I don't think this is true. It just needs permissions to change ownership on the `/data/packages` volume, right?

> Using the docker image for anything else than docker run (this currently also throws an error in the docker tests (when the script calls docker run pypiserver:test --help , but this error is ignored)

Yep this is definitely a problem.

> It uses the default wsgiref server which is non-theaded and not production ready. We should pick another.

In general people should be using the docker image like the do the application, which would mean putting it behind a more appropriate nginx or similar ingress for production workloads.

> I noticed that the pypiserver uses /data as a root, but /data/packages as a volume, which means that new package uploads may not be persisted

`/data` is used as the WORKDIR and the `$HOME` for the pypiserver user, and `/data/packages` is declared as a VOLUME. I'm pretty sure that anything declared as a VOLUME will be persisted unless the container is run with `--rm` or unless the volumes are explicitly removed.

> Currently PORT is used as an environment variable to customize the port. While environment variables for configuration of containers is considered good practice, I believe we should make this variable more specifc, ie PYPISERVER_PORT. We can then also expand this to other configuration settings, and maybe not limit it to only the docker image.

This seems like a good idea.

> I think the entrypoint.sh currently does more harm than good and we should try and remove it again (while keeping docker-compose working), or improve on it drastically

I agree with this. As I mentioned in the other thread, #330 has been I think more more pain than gain.

**Comment by elfjes** [comment]
> > Requires root permissions (which leads to #357)
> 
> I don't think this is true. It just needs permissions to change ownership on the `/data/packages` volume, right?

Well currently it needs to run `chown` and `gosu`, but technically you're right, we could probably give the docker user the correct permissions to do that

> 
> > Using the docker image for anything else than docker run (this currently also throws an error in the docker tests (when the script calls docker run pypiserver:test --help , but this error is ignored)
> 
> Yep this is definitely a problem.
> 
> > It uses the default wsgiref server which is non-theaded and not production ready. We should pick another.
> 
> In general people should be using the docker image like the do the application, which would mean putting it behind a more appropriate nginx or similar ingress for production workloads.

definitely, but this doesn't solve the non-threadedness, and some lockups that some users are experiencing with the default server (probably also due to being single-threaded)
> 
> > I noticed that the pypiserver uses /data as a root, but /data/packages as a volume, which means that new package uploads may not be persisted
> 
> `/data` is used as the WORKDIR and the `$HOME` for the pypiserver user, and `/data/packages` is declared as a VOLUME. I'm pretty sure that anything declared as a VOLUME will be persisted unless the container is run with `--rm` or unless the volumes are explicitly removed.

The package roots defaults to the user `$HOME`, currently `/data`. pypiserver picks up packages that are in `/data/packages` because this is a subdir of `/data`. However, when uploading a package, this will be placed in the root `/data` directly, which is not part of the volume. I believe this is a bug that I introduced when I tried to fix `entrypoint.sh`
> 
> > Currently PORT is used as an environment variable to customize the port. While environment variables for configuration of containers is considered good practice, I believe we should make this variable more specifc, ie PYPISERVER_PORT. We can then also expand this to other configuration settings, and maybe not limit it to only the docker image.
> 
> This seems like a good idea.
> 
> > I think the entrypoint.sh currently does more harm than good and we should try and remove it again (while keeping docker-compose working), or improve on it drastically
> 
> I agree with this. As I mentioned in the other thread, #330 has been I think more more pain than gain.

**Comment by mplanchard** [comment]
> Well currently it needs to run chown and gosu, but technically you're right, we could probably give the docker user the correct permissions to do that

Yeah, the permissions thing is altogether a pain, particularly b/c of the differences between the way the docker daemon works on mac and linux. #256 has a good rundown.

>  definitely, but this doesn't solve the non-threadedness, and some lockups that some users are experiencing with the default server (probably also due to being single-threaded)

Ah yeah that is true. We should probably run it through gunicorn or something at least then

> However, when uploading a package, this will be placed in the root /data directly, which is not part of the volume

Oh, I see. Are we sure this is happening? Our default package directory is ~/packages, and since the Docker user's home is `/data`, it seems like it should be putting packages into `/data/packages`

**Comment by elfjes** [comment]
> > Well currently it needs to run chown and gosu, but technically you're right, we could probably give the docker user the correct permissions to do that
> 
> Yeah, the permissions thing is altogether a pain, particularly b/c of the differences between the way the docker daemon works on mac and linux. #256 has a good rundown.

I just went through #256, which does make some choices more clear. Let's keep the discussion of this issue in the other thread maybe #357?.

> 
> > definitely, but this doesn't solve the non-threadedness, and some lockups that some users are experiencing with the default server (probably also due to being single-threaded)
> 
> Ah yeah that is true. We should probably run it through gunicorn or something at least then

I think gunicorn may not be the best option though, since we'd need to invoke it directly rather than through `pypi-server run --server`, which would then make it harder to change other configuration settings or run the docker image with `pypiserver update`. 
> 
> > However, when uploading a package, this will be placed in the root /data directly, which is not part of the volume
> 
> Oh, I see. Are we sure this is happening? Our default package directory is ~/packages, and since the Docker user's home is `/data`, it seems like it should be putting packages into `/data/packages`

ah, yes you are right, default root is `~/packages`, so all should be fine then. false alarm :)
